File: src/main/java/com/zynchub/digital/hubservice/app/controller/AdminObservabilityController.java
package com.zynchub.digital.hubservice.app.controller;

import com.zynchub.digital.hubservice.app.common.ApiPaths;
import com.zynchub.digital.hubservice.app.common.response.ApiResponse;
import com.zynchub.digital.hubservice.app.dto.AdminObservabilityOverviewResponseDto;
import com.zynchub.digital.hubservice.app.dto.ConnectorHealthListResponseDto;
import com.zynchub.digital.hubservice.app.dto.TenantHealthListResponseDto;
import com.zynchub.digital.hubservice.app.service.ObservabilityService;
import java.time.Instant;
import java.util.Set;
import org.springframework.http.HttpStatus;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.server.ResponseStatusException;

@RestController
@RequestMapping(ApiPaths.API_V3 + "/admin/observability")
public class AdminObservabilityController {

    private static final Set<String> ADMIN_AUTHORITIES = Set.of(
            "ROLE_ADMIN",
            "ROLE_AXIS_ADMIN",
            "ROLE_SUPER_ADMIN",
            "SCOPE_admin",
            "SCOPE_axis.admin");

    private final ObservabilityService observabilityService;

    public AdminObservabilityController(ObservabilityService observabilityService) {
        this.observabilityService = observabilityService;
    }

    @GetMapping("/overview")
    public ApiResponse<AdminObservabilityOverviewResponseDto> getOverview(
            @RequestParam(value = "from", required = false) Instant from,
            @RequestParam(value = "to", required = false) Instant to,
            @RequestParam(value = "periodSeconds", defaultValue = "300") int periodSeconds,
            @RequestParam(value = "tenantId", required = false) String tenantId,
            @RequestParam(value = "connectorId", required = false) String connectorId,
            Authentication authentication) {
        ensureAdmin(authentication);
        return ApiResponse.ok(
                observabilityService.getAdminOverview(from, to, periodSeconds, tenantId, connectorId));
    }

    @GetMapping("/tenants")
    public ApiResponse<TenantHealthListResponseDto> getTenantHealth(
            @RequestParam(value = "from", required = false) Instant from,
            @RequestParam(value = "to", required = false) Instant to,
            @RequestParam(value = "status", required = false) String status,
            @RequestParam(value = "page", defaultValue = "1") int page,
            @RequestParam(value = "size", defaultValue = "50") int size,
            Authentication authentication) {
        ensureAdmin(authentication);
        return ApiResponse.ok(observabilityService.getAdminTenantHealth(from, to, status, page, size));
    }

    @GetMapping("/connectors")
    public ApiResponse<ConnectorHealthListResponseDto> getConnectorHealth(
            @RequestParam(value = "from", required = false) Instant from,
            @RequestParam(value = "to", required = false) Instant to,
            @RequestParam(value = "status", required = false) String status,
            @RequestParam(value = "page", defaultValue = "1") int page,
            @RequestParam(value = "size", defaultValue = "50") int size,
            Authentication authentication) {
        ensureAdmin(authentication);
        return ApiResponse.ok(observabilityService.getAdminConnectorHealth(from, to, status, page, size));
    }

    private void ensureAdmin(Authentication authentication) {
        if (authentication == null) {
            return;
        }
        boolean authorized = authentication.getAuthorities().stream()
                .map(grantedAuthority -> grantedAuthority == null ? null : grantedAuthority.getAuthority())
                .anyMatch(ADMIN_AUTHORITIES::contains);
        if (!authorized) {
            throw new ResponseStatusException(HttpStatus.FORBIDDEN, "Admin role required");
        }
    }
}

File: internal/app/controller/admin_observability_controller.go
package controller

import (
	"net/http"

	"plexus-bff-service-go/internal/app/apiresponse"
	"plexus-bff-service-go/internal/app/security"
	"plexus-bff-service-go/internal/app/service"
)

type AdminObservabilityController struct {
	service service.ObservabilityService
}

func NewAdminObservabilityController(observabilityService service.ObservabilityService) *AdminObservabilityController {
	return &AdminObservabilityController{service: observabilityService}
}

func (c *AdminObservabilityController) Overview(w http.ResponseWriter, r *http.Request) {
	if !ensureAdmin(r) {
		writeJSON(w, http.StatusForbidden, apiresponse.Error("Admin role required"))
		return
	}
	from, to := parseTimeRange(r)
	response, err := c.service.GetAdminOverview(
		r.Context(),
		from,
		to,
		parseIntDefault(r.URL.Query().Get("periodSeconds"), 300),
		r.URL.Query().Get("tenantId"),
		r.URL.Query().Get("connectorId"),
	)
	if err != nil {
		writeJSON(w, http.StatusBadGateway, apiresponse.Error(err.Error()))
		return
	}
	writeJSON(w, http.StatusOK, apiresponse.OK(response))
}

func (c *AdminObservabilityController) Tenants(w http.ResponseWriter, r *http.Request) {
	if !ensureAdmin(r) {
		writeJSON(w, http.StatusForbidden, apiresponse.Error("Admin role required"))
		return
	}
	from, to := parseTimeRange(r)
	response, err := c.service.GetAdminTenantHealth(
		r.Context(),
		from,
		to,
		r.URL.Query().Get("status"),
		parseIntDefault(r.URL.Query().Get("page"), 1),
		parseIntDefault(r.URL.Query().Get("size"), 50),
	)
	if err != nil {
		writeJSON(w, http.StatusBadGateway, apiresponse.Error(err.Error()))
		return
	}
	writeJSON(w, http.StatusOK, apiresponse.OK(response))
}

func (c *AdminObservabilityController) Connectors(w http.ResponseWriter, r *http.Request) {
	if !ensureAdmin(r) {
		writeJSON(w, http.StatusForbidden, apiresponse.Error("Admin role required"))
		return
	}
	from, to := parseTimeRange(r)
	response, err := c.service.GetAdminConnectorHealth(
		r.Context(),
		from,
		to,
		r.URL.Query().Get("status"),
		parseIntDefault(r.URL.Query().Get("page"), 1),
		parseIntDefault(r.URL.Query().Get("size"), 50),
	)
	if err != nil {
		writeJSON(w, http.StatusBadGateway, apiresponse.Error(err.Error()))
		return
	}
	writeJSON(w, http.StatusOK, apiresponse.OK(response))
}

func ensureAdmin(r *http.Request) bool {
	auth := security.FromContext(r.Context())
	if auth == nil {
		return true
	}
	adminAuthorities := map[string]struct{}{
		"ROLE_ADMIN":      {},
		"ROLE_AXIS_ADMIN": {},
		"ROLE_SUPER_ADMIN": {},
		"SCOPE_admin":     {},
		"SCOPE_axis.admin": {},
	}
	for _, authority := range auth.Authorities {
		if _, ok := adminAuthorities[authority]; ok {
			return true
		}
	}
	return false
}
